packages/tpetra/core/test/MultiVector/ForEach.cpp

```diff
@@ -0,0 +1,423 @@
+
+    RCP<Teuchos::FancyOStream> outPtr = debug ?
+      Teuchos::getFancyOStream (Teuchos::rcpFromRef (std::cerr)) :
+      Teuchos::rcpFromRef (out);
+    Teuchos::FancyOStream& myOut = *outPtr;
+
+    myOut << "Test Tpetra::for_each" << endl;
+    Teuchos::OSTab tab0 (myOut);
+
+    myOut << "Create a Map" << endl;
+    auto comm = getDefaultComm ();
+    const auto INVALID = Teuchos::OrdinalTraits<GST>::invalid ();
+    const size_t numLocal = 13;
+    const size_t numVecs  = 3;
+    const GO indexBase = 0;
+    auto map = rcp (new map_type (INVALID, numLocal, indexBase, comm));
+
+    myOut << "Create a MultiVector, and make sure that it has "
+      "the right number of vectors (columns)" << endl;
+    multivec_type X (map, numVecs);
+    TEST_EQUALITY( X.getNumVectors (), numVecs );
+
+    out << "Test for_each(MV, void(double&)): Set entries to 418" << endl;
+    for_each ("X_ij=418", X, KOKKOS_LAMBDA (double& X_ij) {
+        X_ij = 418.0;
+      });
+    {
+      X.sync_host ();
+      auto X_lcl = X.getLocalViewHost ();
+      for (LO j = 0; j < LO (X.getNumVectors ()); ++j) {
+        out << "Column " << j << std::endl;
+        bool ok = true;
+        for (LO i = 0; i < LO (X.getLocalLength ()); ++i) {
+          const double expectedVal = 418.0;
+          if (X_lcl(i,j) != expectedVal) {
+            out << "X_lcl(" << i << "," << j << ") = " << X_lcl(i,j)
+                << " != " << expectedVal << endl;
+            ok = false;
+          }
+        }
+        TEST_ASSERT( ok );
+      }
+    }
+
+    lclSuccess = success ? 1 : 0;
+    reduceAll<int, int> (*comm, REDUCE_MIN, lclSuccess, outArg (gblSuccess));
+    TEST_ASSERT( gblSuccess == 1 );
+    if (gblSuccess != 1) {
+      out << "Returning early" << endl;
+      return;
+    }
+
+    out << "Test for_each(DefaultHostExecutionSpace, MV, "
+      "void(double&)): Set entries to 777" << endl;
+    for_each ("X_ij=777", Kokkos::DefaultHostExecutionSpace (), X,
+              KOKKOS_LAMBDA (double& X_ij) {
+                X_ij = 777.0;
+              });
+    {
+      //X.sync_host ();
+      auto X_lcl = X.getLocalViewHost ();
+      for (LO j = 0; j < LO (X.getNumVectors ()); ++j) {
+        out << "Column " << j << std::endl;
+        bool ok = true;
+        for (LO i = 0; i < LO (X.getLocalLength ()); ++i) {
+          const double expectedVal = 777.0;
+          if (X_lcl(i,j) != expectedVal) {
+            out << "X_lcl(" << i << "," << j << ") = " << X_lcl(i,j)
+                << " != " << expectedVal << std::endl;
+            ok = false;
+          }
+        }
+        TEST_ASSERT( ok );
+      }
+    }
+
+    lclSuccess = success ? 1 : 0;
+    reduceAll<int, int> (*comm, REDUCE_MIN, lclSuccess, outArg (gblSuccess));
+    TEST_ASSERT( gblSuccess == 1 );
+    if (gblSuccess != 1) {
+      out << "Returning early" << endl;
+      return;
+    }
+
+    out << "X.sync_device(); X.putScalar(666.0);" << endl;
+    X.sync_device ();
+    X.putScalar (666.0);
```

Comment:
I see what you did there...

Reply:
"It's a solar number.  You may call me 'Little Sunshine.'"